xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -615,7 +630,90 @@ unsigned int CGUIFontTTF::GetTextureLineHeight() const
   return m_cellHeight + spacing_between_characters_in_texture;
 }
 
-CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr)
+vecGlyphInfo CGUIFontTTF::GetHarfbuzzShapedGlyphs(const vecText &text)
+{
+    vecGlyphInfo g_infos;
+    if(text.empty())
+    {
+      return g_infos;
+    }
+    std::vector<hb_script_t> t_scripts;
+    std::vector<RunInfo> t_runs;
+    hb_unicode_funcs_t *ufuncs = hb_unicode_funcs_create( hb_unicode_funcs_get_default() );
+    hb_script_t last_script;
+    int last_script_index = -1;
+    int last_set_index = -1;
+
+    for (unsigned int i = 0; i < text.size(); i++)
+    {
+      t_scripts.push_back(hb_unicode_script( ufuncs, (wchar_t)(0xffff & text[i]) ));
+    }
+    // HB_SCRIPT_COMMON or HB_SCRIPT_INHERITED should be replaced with previous script
+    for (unsigned int i = 0; i < t_scripts.size(); i++)
+    {
+      if (t_scripts[i] == HB_SCRIPT_COMMON || t_scripts[i] == HB_SCRIPT_INHERITED)
+      {
+        if (last_script_index != -1)
+        {
+          t_scripts[i] = last_script;
+          last_set_index = i;
+        }
+      }
+      else
+      {
+        for (int j = last_set_index + 1; j < i; j++)
+          t_scripts[j] = t_scripts[i];
+        last_script = t_scripts[i];
+        last_script_index = i;
+        last_set_index = i;
+      }
+    }
+    last_script = t_scripts[0];
+    int last_run_start = 0;
+    for (unsigned int i = 0; i <=  t_scripts.size(); i++)
+    {
+      if (i == t_scripts.size() || t_scripts[i] != last_script)
+      {
+        RunInfo run;
+        run.start_offset = last_run_start;
+        run.end_offset = i;
+        run.script = last_script;
+        t_runs.push_back(run);
+        if (i < t_scripts.size())
+        {
+            last_script = t_scripts[i];
+            last_run_start = i;
+        }
+        else
+        {
+            break;
+        }
+      }
+    }
+    for (unsigned int i = 0; i < t_runs.size(); i++)
+    {
+      t_runs[i].buffer = hb_buffer_create();
+      hb_buffer_set_direction(t_runs[i].buffer, (hb_direction_t)HB_DIRECTION_LTR);
```

Comment:
I think this should be correctly set to HB_DIRECTION_RTL for languages like Arabic, Hebrew and others?

Reply:
Is it already taken care of by FriBidi in https://github.com/xbmc/xbmc/blob/master/xbmc/utils/CharsetConverter.cpp ?. Am I correct?

Reply:
This PR only changes the current kodi behaviour of text->characters->glyphs to text->shaping->glyphs

Reply:
Seems you are right. I  tried same settings on a different system and it looks the same.